webapp/src/gitjson.tsx

```diff
@@ -0,0 +1,693 @@
+                            <span className="files-to-restore">{deletedFiles.join(", ")}</span>
+                        </p>}
+                    {addedFiles.length == 0 ? undefined :
+                        <p>
+                            {lf("Reverting this file will also remove:")}
+                            {" "}
+                            <span className="files-to-delete">{addedFiles.join(", ")}</span>
+                        </p>}
+                </div>
+                {isBlocks ? <div className="ui segment"><p>{lf("Some blocks changed")}</p></div> : diffJSX.length ?
+                    <div className="ui segment diff">
+                        <table className="diffview">
+                            <tbody>
+                                {diffJSX}
+                            </tbody>
+                        </table>
+                    </div>
+                    :
+                    <div className="ui segment">
+                        <p>{lf("Whitespace changes only.")}</p>
+                    </div>
+                }
+            </div>)
+        return cache.diff
+    }
+
+
+    display() {
+        if (!this.isVisible)
+            return undefined;
+
+        const header = this.parent.state.header;
+        if (!header || !header.githubId) return undefined;
+        // TODO: disable commit changes if no changes available
+        // TODO: commitAsync handle missing token or failed push
+
+        const diffFiles = pkg.mainEditorPkg().sortedFiles().filter(p => p.baseGitContent != p.content)
+        const needsCommit = diffFiles.length > 0
+
+        if (this.needsPull == null) {
+            this.needsPull = true
+            workspace.hasPullAsync(this.parent.state.header)
+                .then(v => {
+                    if (v != this.needsPull) {
+                        this.needsPull = v
+                        this.parent.setState({})
+                    }
+                })
+                .catch(this.handleGithubError)
+        }
+
+        /**
+         *                                     <sui.PlainCheckbox
+                                        label={lf("Publish to users (increment version)")}
+                                        onChange={this.setBump} />
+    
+         */
+        const githubId = this.parsedRepoId()
+        const master = githubId.tag == "master";
+        const gs = this.getGitJson();
+        // don't use gs.prUrl, as it gets cleared often
+        const url = `https://github.com/${githubId.fullName}${master ? "" : `/tree/${githubId.tag}`}`;
+        const needsToken = !pxt.github.token;
+        // this will show existing PR if any
+        const prUrl = !gs.isFork && master ? null :
+            `https://github.com/${githubId.fullName}/compare/${githubId.tag}?expand=1`
+        this.needsLicenseMessage = gs.commit && !gs.commit.tree.tree.some(f =>
+            /^LICENSE/.test(f.path.toUpperCase()) || /^COPYING/.test(f.path.toUpperCase()))
+        return (
```

Comment:
I think this needs to be broken up into several React components

Reply:
this one does not need to be in the state, made it a local var.